libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+        auto add_branch_impl(B&& branch, idx_type left_child, idx_type right_child) -> idx_type;
+    };
+
+    /**
+     * A parser for postfix expressions.
+     *
+     * The parser creates an expression tree and validate that the expression being pushed
+     * is a valid postfix expression.
+     * For example, for the expression ``a + b * c`` on might push ``a b c * +``
+     * or ``b c * a +``.
+     */
+    template <typename Variable, typename Operator>
+    class PostfixParser
+    {
+    public:
+
+        using operator_type = Operator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+
+        void push_variable(const variable_type& var);
+        void push_variable(variable_type&& var);
+        void push_operator(const operator_type& op);
+        void push_operator(operator_type&& op);
+        void finalize();
+
+        [[nodiscard]] auto tree() const& -> const tree_type&;
+        [[nodiscard]] auto tree() && -> tree_type&&;
+
+    private:
+
+        using idx_type = typename tree_type::idx_type;
+        using node_idx_stack = std::vector<idx_type>;
+
+        /** The expression tree containing the expression being parsed. */
+        tree_type m_tree = {};
+        /** Orphan nodes are node without a parent. */
+        node_idx_stack m_orphans = {};
+
+        void orphans_push(idx_type idx);
+        auto orphans_pop() -> idx_type;
+
+        template <typename V>
+        void push_variable_impl(V&& var);
+        template <typename O>
+        void push_operator_impl(O&& op);
+    };
+
+    /**
+     * A parser for infix expressions.
+     *
+     * The parser creates an expression tree and validate that the expression being pushed
+     * is a valid infix expression.
+     * For example, the expression ``a + b * c`` can be pushed directly (thanks to the
+     * operator precedence), or parenthesised as ``a + (b * c)``.
+     */
+    template <typename Variable, typename Operator, typename OperatorCmp = std::less<>>
+    class InfixParser
+    {
+    public:
+
+        using operator_type = Operator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+        using operator_precedence_type = OperatorCmp;
+
+        InfixParser(const operator_precedence_type& cmp);
+        InfixParser(operator_precedence_type&& cmp = {});
+
+        void push_variable(const variable_type& var);
+        void push_variable(variable_type&& var);
+        void push_operator(const operator_type& op);
+        void push_operator(operator_type&& op);
+        void push_left_parenthesis();
+        void push_right_parenthesis();
+        void finalize();
+
+        [[nodiscard]] auto tree() const& -> const tree_type&;
+        [[nodiscard]] auto tree() && -> tree_type&&;
+
+    private:
+
+        using postfix_parser_type = PostfixParser<variable_type, operator_type>;
+        struct LeftParenthesis
```

Comment:
does the style forces you multiple lines here? to me that would be better as a one-liner
but if the style is like that dont bother changing that.

Reply:
I agree, but yes clang-format does it that way. Cannot find an option to configure it...